libraries/matrix/impl/src/main/kotlin/io/element/android/libraries/matrix/impl/encryption/RustEncryptionService.kt

```diff
@@ -0,0 +1,133 @@
+import org.matrix.rustcomponents.sdk.Encryption
+import org.matrix.rustcomponents.sdk.RecoveryStateListener
+import org.matrix.rustcomponents.sdk.BackupState as RustBackupState
+import org.matrix.rustcomponents.sdk.BackupUploadState as RustBackupUploadState
+import org.matrix.rustcomponents.sdk.EnableRecoveryProgress as RustEnableRecoveryProgress
+import org.matrix.rustcomponents.sdk.RecoveryState as RustRecoveryState
+
+internal class RustEncryptionService(
+    client: Client,
+    private val dispatchers: CoroutineDispatchers,
+) : EncryptionService {
+
+    private val service: Encryption = client.encryption()
+
+    private val backupStateMapper = BackupStateMapper()
+    private val recoveryStateMapper = RecoveryStateMapper()
+    private val enableRecoveryProgressMapper = EnableRecoveryProgressMapper()
+    private val backupUploadStateMapper = BackupUploadStateMapper()
+
+    override val backupStateStateFlow: MutableStateFlow<BackupState> = MutableStateFlow(service.backupState().let(backupStateMapper::map))
+    override val recoveryStateStateFlow: MutableStateFlow<RecoveryState> = MutableStateFlow(service.recoveryState().let(recoveryStateMapper::map))
+    override val enableRecoveryProgressStateFlow: MutableStateFlow<EnableRecoveryProgress> = MutableStateFlow(EnableRecoveryProgress.Unknown)
+    override val backupUploadStateStateFlow: MutableStateFlow<BackupUploadState> = MutableStateFlow(BackupUploadState.Unknown)
+
+    fun start() {
+        service.backupStateListener(object : BackupStateListener {
+            override fun onUpdate(status: RustBackupState) {
+                backupStateStateFlow.value = backupStateMapper.map(status)
+            }
+        })
+
+        service.recoveryStateListener(object : RecoveryStateListener {
+            override fun onUpdate(status: RustRecoveryState) {
+                recoveryStateStateFlow.value = recoveryStateMapper.map(status)
+            }
+        })
+    }
+
+    fun destroy() {
+        // No way to remove the listeners...
+        service.destroy()
+    }
+
+    override suspend fun enableBackups(): Result<Unit> = withContext(dispatchers.io) {
+        runCatching {
+            service.enableBackups()
+        }
+    }
+
+    override suspend fun enableRecovery(
+        waitForBackupsToUpload: Boolean,
+    ): Result<Unit> = withContext(dispatchers.io) {
+        runCatching {
+            service.enableRecovery(
+                waitForBackupsToUpload = waitForBackupsToUpload,
+                progressListener = object : EnableRecoveryProgressListener {
+                    override fun onUpdate(status: RustEnableRecoveryProgress) {
+                        enableRecoveryProgressStateFlow.value = enableRecoveryProgressMapper.map(status)
+                    }
+                }
+            )
+                // enableRecovery returns the encryption key, but we read it from the state flow
+                .let { }
```

Comment:
Why `.let{}` when `Unit` was enough ?

Reply:
`Unit` is working, but quality check does not like it.

Reply:
Why doesn't it? It's the most correct thing to do in this case kotlin-wise.

Reply:
Yeah, I agree. Detekt is reporting this:
> ~/workspaces/element-x-android/libraries/matrix/impl/src/main/kotlin/io/element/android/libraries/matrix/impl/encryption/RustEncryptionService.kt:84:13: A single Unit expression is unnecessary and can safely be removed. [OptionalUnit]

There are several similar reports in the detekt project https://github.com/detekt/detekt, but all the issues seems to be close.

In the mean time, using `.let { }` is the solution, and it's already used in the codebase.

Reply:
Seems like a detekt bug